File: rxbinding/src/main/java/com/jakewharton/rxbinding2/view/ViewHoverObservable.java
package com.jakewharton.rxbinding2.view;

import android.view.MotionEvent;
import android.view.View;
import android.view.View.OnHoverListener;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;
import io.reactivex.functions.Predicate;

import static io.reactivex.android.MainThreadDisposable.verifyMainThread;

final class ViewHoverObservable extends Observable<MotionEvent> {
  private final View view;
  private final Predicate<? super MotionEvent> handled;

  ViewHoverObservable(View view, Predicate<? super MotionEvent> handled) {
    this.view = view;
    this.handled = handled;
  }

  @Override protected void subscribeActual(Observer<? super MotionEvent> observer) {
    verifyMainThread();
    Listener listener = new Listener(view, handled, observer);
    observer.onSubscribe(listener);
    view.setOnHoverListener(listener);
  }

  static final class Listener extends MainThreadDisposable implements OnHoverListener {
    private final View view;
    private final Predicate<? super MotionEvent> handled;
    private final Observer<? super MotionEvent> observer;

    Listener(View view, Predicate<? super MotionEvent> handled,
        Observer<? super MotionEvent> observer) {
      this.view = view;
      this.handled = handled;
      this.observer = observer;
    }

    @Override public boolean onHover(View v, MotionEvent event) {
      if (!isDisposed()) {
        try {
          if (handled.test(event)) {
            observer.onNext(event);
            return true;
          }
        } catch (Exception e) {
          observer.onError(e);
          dispose();
        }
      }
      return false;
    }

    @Override protected void onDispose() {
      view.setOnHoverListener(null);
    }
  }
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding3/view/ViewHoverObservable.kt
@file:JvmName("RxView")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.view

import android.view.MotionEvent
import android.view.View
import android.view.View.OnHoverListener
import androidx.annotation.CheckResult
import com.jakewharton.rxbinding3.internal.AlwaysTrue
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable
import io.reactivex.functions.Predicate

import com.jakewharton.rxbinding3.internal.checkMainThread

/**
 * Create an observable of hover events for `view`.
 *
 * *Warning:* Values emitted by this observable are **mutable** and part of a shared
 * object pool and thus are **not safe** to cache or delay reading (such as by observing
 * on a different thread). If you want to cache or delay reading the items emitted then you must
 * map values through a function which calls [MotionEvent.obtain] or
 * [MotionEvent.obtainNoHistory] to create a copy.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Warning:* The created observable uses [View.setOnHoverListener] to observe
 * touches. Only one observable can be used for a view at a time.
 *
 * @param handled Predicate invoked with each value to determine the return value of the
 * underlying [View.OnHoverListener].
 */
@CheckResult
@JvmOverloads
fun View.hovers(
  handled: Predicate<in MotionEvent> = AlwaysTrue
): Observable<MotionEvent> {
  return ViewHoverObservable(this, handled)
}

private class ViewHoverObservable(
  private val view: View,
  private val handled: Predicate<in MotionEvent>
) : Observable<MotionEvent>() {

  override fun subscribeActual(observer: Observer<in MotionEvent>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, handled, observer)
    observer.onSubscribe(listener)
    view.setOnHoverListener(listener)
  }

  private class Listener(
    private val view: View, private val handled: Predicate<in MotionEvent>,
    private val observer: Observer<in MotionEvent>
  ) : MainThreadDisposable(), OnHoverListener {

    override fun onHover(v: View, event: MotionEvent): Boolean {
      if (!isDisposed) {
        try {
          if (handled.test(event)) {
            observer.onNext(event)
            return true
          }
        } catch (e: Exception) {
          observer.onError(e)
          dispose()
        }

      }
      return false
    }

    override fun onDispose() {
      view.setOnHoverListener(null)
    }
  }
}
